lib/dry/cli/command.rb

```diff
@@ -26,33 +25,62 @@ def self.extended(base)
           super
 
           base.class_eval do
+            @_mutex       = Mutex.new
             @description  = nil
-            @examples     = Concurrent::Array.new
-            @subcommands  = Concurrent::Array.new
-            @arguments    = Concurrent::Array.new
-            @options      = Concurrent::Array.new
+            @examples     = []
+            @arguments    = []
+            @options      = []
+            @subcommands  = []
           end
         end
 
         # @since 0.1.0
         # @api private
-        attr_reader :description
+        def description
+          @_mutex.synchronize do
```

Comment:
what is the reason to have synchonize for getters? I thought we only need to sync and lock structure at writing/appending time.

Reply:
@IvanShamatov @flash-gordon The reason why I wrapped getters isn't for atomicity of the operation (it is atomic indeed), but because this reference is hold at the class level and it can be theoretically be altered like this:

```ruby
class Command < Dry::CLI::Command
  example "foo", "bar"
end

Command.examples.reverse!
puts Command.examples # => ["bar", "foo"]
```

Because we can't provide proper encapsulation for this, at least a lock should guarantee thread-safety.

See https://books.google.it/books?id=mzgFCAAAQBAJ&pg=PA59&lpg=PA59&dq=information+escape+thread+safety&source=bl&ots=7W9G821iaY&sig=ACfU3U0ZOeOsLvGLSLoFnfwiQwS-6xhAAw&hl=en&sa=X&ved=2ahUKEwjQhbWayaToAhXTUhUIHU-HAScQ6AEwDHoECA0QAQ#v=onepage&q=information%20escape%20thread%20safety&f=false

WDYT?

Reply:
@jodosha since the mutex is attached to a command instance and not to its ivars, it won't actually help. Once `.examples` returns the array can be mutated at will, even in a thread-unsafe manner. If you want to secure it there are two ways I can think of
1. `.dup` values returned by `examples`, i.e. 
  ```ruby
  def examples
    @examples.dup
  end
  ```
2. `.freeze` values every time
  ```ruby
  def example(*examples)
    @_mutex.synchronize do
      @examples = (@examples + examples.flatten(1)).freeze
    end
  end
  ```

Reply:
@flash-gordon Yeah, on a second thought, my point isn't about thread-safety, but mutability of the state.

I'll remove the mutex around those operations.

Thanks all for the discussion.